gazebo_ros/src/qos.cpp

```diff
@@ -0,0 +1,219 @@
+      qos_overrides.reliability = RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT;
+    } else if ("reliable" == reliability) {
+      qos_overrides.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+    } else if ("best_effort" == reliability) {
+      qos_overrides.reliability = RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT;
+    } else {
+      std::stringstream oss;
+      oss << "invalid setting for reliability '" << reliability << "'";
+      throw InvalidQoSException(oss.str());
+    }
+  }
+
+  // Parse 'durability' QoS
+  if (_sdf->HasElement("durability")) {
+    auto durability = _sdf->GetElement("durability")->Get<std::string>();
+    if ("system" == durability) {
+      qos_overrides.durability = RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT;
+    } else if ("volatile" == durability) {
+      qos_overrides.durability = RMW_QOS_POLICY_DURABILITY_VOLATILE;
+    } else if ("transient_local" == durability) {
+      qos_overrides.durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
+    } else {
+      std::stringstream oss;
+      oss << "invalid setting for durability '" << durability << "'";
+      throw InvalidQoSException(oss.str());
+    }
+  }
+
+  // Parse 'history' QoS
+  if (_sdf->HasElement("history")) {
+    auto history_element = _sdf->GetElement("history");
+    auto history = history_element->Get<std::string>();
+    if ("system" == history) {
+      qos_overrides.history = RMW_QOS_POLICY_HISTORY_SYSTEM_DEFAULT;
+    } else if ("keep_last" == history) {
+      if (!history_element->HasAttribute("depth")) {
+        throw InvalidQoSException("'keep_last' used without providing a depth");
+      }
+      auto depth = history_element->Get<size_t>("depth");
+      qos_overrides.history = RMW_QOS_POLICY_HISTORY_KEEP_LAST;
+      qos_overrides.depth = depth;
+    } else if ("keep_all" == history) {
+      qos_overrides.history = RMW_QOS_POLICY_HISTORY_KEEP_ALL;
+    } else {
+      std::stringstream oss;
+      oss << "invalid setting for history '" << history << "'";
+      throw InvalidQoSException(oss.str());
+    }
+  }
+
+  // Parse 'deadline' QoS
+  if (_sdf->HasElement("deadline")) {
+    auto deadline = _sdf->GetElement("deadline")->Get<size_t>();
+    qos_overrides.deadline = std::chrono::milliseconds(deadline);
+  }
+
+  // Parse 'lifespan' QoS
+  if (_sdf->HasElement("lifespan")) {
+    auto lifespan = _sdf->GetElement("lifespan")->Get<size_t>();
+    qos_overrides.lifespan = std::chrono::milliseconds(lifespan);
+  }
+
+  // Parse 'liveliness' QoS
+  if (_sdf->HasElement("liveliness")) {
+    auto liveliness = _sdf->GetElement("liveliness")->Get<std::string>();
+    if ("system" == liveliness) {
+      qos_overrides.liveliness = RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT;
+    } else if ("automatic" == liveliness) {
+      qos_overrides.liveliness = RMW_QOS_POLICY_LIVELINESS_AUTOMATIC;
+    } else if ("manual_by_node" == liveliness) {
+      qos_overrides.liveliness = RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE;
```

Comment:
I don't know why, but should we not include it here then?

Reply:
That feature depended on participants being mapped to nodes, and there was agreement to remove it.
I will create a couple of PRs to actually remove it, but it can already be removed from here.

Reply:
Removed in b52f8c2